File: iOS/Retainic/ReviewOrdering.swift
//
//  ReviewOrdering.swift
//  Retainic
//
//  Lapse-derived weighted ordering for practice sessions. This intentionally
//  uses the aggregate memoryStats already persisted by every client.
//

import Foundation

extension VocabWord {
    func memoryLapses(for aspect: String) -> Int {
        guard let stat = memoryStats?[aspect] else { return 0 }
        return max(0, stat.seen - stat.timesRemembered)
    }

    func reviewWeight(for aspect: String) -> Double {
        Double(1 + min(memoryLapses(for: aspect), 9))
    }
}

/// Weighted random permutation using exponential-race keys. Every item remains
/// in the deck exactly once; higher-weight items merely tend to sort earlier.
func weightedReviewOrder<Item>(
    _ items: [Item],
    weight: (Item) -> Double,
    random: () -> Double = { Double.random(in: 0..<1) }
) -> [Item] {
    var keyed: [(item: Item, index: Int, key: Double)] = []
    keyed.reserveCapacity(items.count)
    for (index, item) in items.enumerated() {
        let boundedWeight = max(1, weight(item))
        let unit = min(1 - Double.ulpOfOne, max(Double.leastNonzeroMagnitude, random()))
        keyed.append((item, index, -log(unit) / boundedWeight))
    }
    keyed.sort { lhs, rhs in
        lhs.key == rhs.key ? lhs.index < rhs.index : lhs.key < rhs.key
    }
    return keyed.map(\.item)
}

File: android/app/src/main/java/com/retainic/app/ui/ReviewOrdering.kt
package com.retainic.app.ui

import com.retainic.app.data.VocabWord
import kotlin.math.ln
import kotlin.math.max
import kotlin.math.min
import kotlin.random.Random

/** Incorrect recalls already persisted for one aspect. */
internal fun VocabWord.memoryLapses(aspect: String): Int {
    val stat = memoryStats?.get(aspect) ?: return 0
    return max(0, stat.seen - stat.timesRemembered)
}

/** Bounded priority multiplier: 1 for unseen/perfect items, up to 10. */
internal fun VocabWord.reviewWeight(aspect: String): Double =
    (1 + min(memoryLapses(aspect), 9)).toDouble()

/** Weighted random permutation using exponential-race keys. */
internal fun <Item> weightedReviewOrder(
    items: List<Item>,
    weight: (Item) -> Double,
    random: () -> Double = { Random.nextDouble() },
): List<Item> = items.mapIndexed { index, item ->
    val boundedWeight = max(1.0, weight(item))
    val unit = random().coerceIn(Double.MIN_VALUE, 1.0 - Math.ulp(1.0))
    Triple(item, index, -ln(unit) / boundedWeight)
}.sortedWith(compareBy<Triple<Item, Int, Double>> { it.third }.thenBy { it.second })
    .map { it.first }
